File: src/reporter/report_generator.py
from datetime import datetime
import uuid
from typing import Dict, Any, Optional, List
from .types import (
    MultiPageReportData,
    TestData,
    PageReport,
)

def parse_multi_page_analysis_results(
    pr_number: str,
    repository: str,
    page_results: List[Dict[str, Any]]
) -> MultiPageReportData:
    """
    Parse multiple page analysis results and generate a structured multi-page report.

    Args:
        pr_number: PR number
        repository: Repository name
        page_results: List of page result dictionaries containing:
            - page_path: str
            - base_url: str
            - pr_url: str
            - visual_analysis: Dict[str, Any]
            - sections_analysis: str
            - image_refs: Optional[Dict[str, str]]

    Returns:
        MultiPageReportData with test_data and reports
    """
    # Create test data
    test_data: TestData = {
        "pr_number": pr_number,
        "repository": repository,
        "timestamp": datetime.utcnow().isoformat()
    }

    # Process each page result
    reports: List[PageReport] = []

    for page_result in page_results:
        page_path = page_result['page_path']
        base_url = page_result['base_url']
        pr_url = page_result['pr_url']
        visual_analysis = page_result['visual_analysis']
        sections_analysis = page_result['sections_analysis']
        image_refs = page_result.get('image_refs')

        # Determine status based on visual analysis
        status = "pass"  # Default to pass
        if isinstance(visual_analysis, dict):
            # Check for critical issues
            critical_issues_enum = visual_analysis.get("critical_issues_enum", "none")
            if critical_issues_enum != "none":
                status = "fail"
            else:
                # Check for visual changes
                visual_changes_enum = visual_analysis.get("visual_changes_enum", "none")
                recommendation_enum = visual_analysis.get("recommendation_enum", "pass")

                if visual_changes_enum == "significant" or recommendation_enum == "review_required":
                    status = "warning"
                elif visual_changes_enum == "minor":
                    status = "warning"
                else:
                    status = "pass"
        else:
            # Fallback for string-based analysis
            analysis_text = str(visual_analysis).lower()
            if "missing sections" in analysis_text or "critical" in analysis_text:
                status = "fail"
            elif ("significant changes" in analysis_text and "no significant changes" not in analysis_text) or "review required" in analysis_text:
                status = "warning"
            else:
                status = "pass"

        # Parse the visual analysis to extract structured data
        # (Assume visual_analysis is already structured for multi-page)
        if isinstance(visual_analysis, dict):
            parsed_report = visual_analysis
        else:
            # Create default structured data for string-based analysis
            parsed_report = {
                "critical_issues": {"sections": [], "summary": ""},
                "structural_analysis": {"section_order": "", "layout": "", "broken_layouts": ""},
                "visual_changes": {"diff_highlights": [], "animation_issues": "", "conclusion": ""},
                "conclusion": {"critical_issues": "", "visual_changes": "", "recommendation": "", "summary": ""}
            }

        # Create page report with guaranteed valid status
        page_report: PageReport = {
            "page_path": page_path,
            "url": base_url,
            "preview_url": pr_url,
            "status": status,  # Use our determined status instead of parsed one
            "critical_issues": parsed_report["critical_issues"],
            "critical_issues_enum": visual_analysis.get("critical_issues_enum", "none") if isinstance(visual_analysis, dict) else ("missing_sections" if "missing sections" in str(visual_analysis).lower() else "other_issues" if "critical" in str(visual_analysis).lower() else "none"),
            "structural_analysis": parsed_report["structural_analysis"],
            "visual_changes": parsed_report["visual_changes"],
            "visual_changes_enum": visual_analysis.get("visual_changes_enum", "none") if isinstance(visual_analysis, dict) else ("significant" if "significant changes" in str(visual_analysis).lower() and "no significant changes" not in str(visual_analysis).lower() else "minor" if "minor" in str(visual_analysis).lower() else "none"),
            "recommendation_enum": visual_analysis.get("recommendation_enum", "pass") if isinstance(visual_analysis, dict) else ("review_required" if "review required" in str(visual_analysis).lower() else "pass"),
            "conclusion": parsed_report["conclusion"],
            "image_refs": image_refs
        }

        reports.append(page_report)

    return {
        "test_data": test_data,
        "reports": reports
    }

File: src/reporter/report-generator.ts
import type {
  MultiPageReportData,
  TestData,
  PageReport,
  ReportStatus,
  CriticalIssuesStatus,
  VisualChangesStatus,
  RecommendationStatus,
  ImageReferences,
} from "./types.js";
import type { VisualAnalysisResult } from "../vision/types.js";

/**
 * Determine status from visual analysis data.
 *
 * @param visualAnalysis - Visual analysis data (dict or string)
 * @returns Status string: 'pass', 'warning', 'fail', or 'none'
 */
export function determineStatusFromVisualAnalysis(
  visualAnalysis: VisualAnalysisResult | Record<string, any> | string | null
): ReportStatus | "none" {
  if (!visualAnalysis) {
    return "none";
  }

  if (
    typeof visualAnalysis === "string" ||
    !(visualAnalysis instanceof Object)
  ) {
    // Fallback for string-based analysis
    const analysisText = String(visualAnalysis).toLowerCase();
    if (
      analysisText.includes("missing sections") ||
      analysisText.includes("critical")
    ) {
      return "fail";
    } else if (
      (analysisText.includes("significant changes") &&
        !analysisText.includes("no significant changes")) ||
      analysisText.includes("review required")
    ) {
      return "warning";
    } else {
      return "pass";
    }
  }

  // Determine status based on visual analysis (original logic)
  let status: ReportStatus | "none" = "pass"; // Default to pass

  const analysis = visualAnalysis as Record<string, any>;

  // Check for critical issues
  const criticalIssuesEnum = analysis.critical_issues_enum || "none";
  if (criticalIssuesEnum !== "none") {
    status = "fail";
  } else {
    // Check for visual changes
    const visualChangesEnum = analysis.visual_changes_enum || "none";
    const recommendationEnum = analysis.recommendation_enum || "pass";

    if (
      visualChangesEnum === "significant" ||
      recommendationEnum === "review_required"
    ) {
      status = "warning";
    } else if (visualChangesEnum === "minor") {
      status = "warning";
    } else {
      status = "pass";
    }
  }

  return status;
}

/**
 * Parse multiple page analysis results and generate a structured multi-page report.
 *
 * @param prNumber - PR number
 * @param repository - Repository name
 * @param pageResults - List of page result objects containing:
 *   - page_path: string
 *   - base_url: string
 *   - pr_url: string
 *   - visual_analysis: VisualAnalysisResult or dict
 *   - sections_analysis: string
 *   - image_refs: Optional ImageReferences
 * @returns MultiPageReportData with test_data and reports
 */
export function parseMultiPageAnalysisResults(
  prNumber: string,
  repository: string,
  pageResults: Array<{
    page_path: string;
    base_url: string;
    pr_url: string;
    visual_analysis: VisualAnalysisResult | Record<string, any>;
    sections_analysis: string;
    image_refs?: ImageReferences | null;
  }>
): MultiPageReportData {
  // Create test data
  const testData: TestData = {
    pr_number: prNumber,
    repository: repository,
    timestamp: new Date().toISOString(),
  };

  // Process each page result
  const reports: PageReport[] = [];

  for (const pageResult of pageResults) {
    const { page_path, base_url, pr_url, visual_analysis, image_refs } =
      pageResult;

    // Determine status using the general function
    console.log("visual_analysis API ------- ", visual_analysis);
    const statusResult = determineStatusFromVisualAnalysis(visual_analysis);
    // Ensure status is ReportStatus (not 'none')
    const status: ReportStatus =
      statusResult === "none" ? "pass" : statusResult;

    // Parse the visual analysis to extract structured data
    // (Assume visual_analysis is already structured for multi-page)
    let parsedReport: Record<string, any>;
    if (
      typeof visual_analysis === "object" &&
      visual_analysis !== null &&
      !Array.isArray(visual_analysis)
    ) {
      parsedReport = visual_analysis as Record<string, any>;
    } else {
      // Create default structured data for string-based analysis
      parsedReport = {
        critical_issues: { sections: [], summary: "" },
        structural_analysis: {
          section_order: "",
          layout: "",
          broken_layouts: "",
        },
        visual_changes: {
          diff_highlights: [],
          animation_issues: "",
          conclusion: "",
        },
        conclusion: {
          critical_issues: "",
          visual_changes: "",
          recommendation: "",
          summary: "",
        },
      };
    }

    // Extract enum values with fallbacks
    const analysisDict =
      typeof visual_analysis === "object" &&
      visual_analysis !== null &&
      !Array.isArray(visual_analysis)
        ? (visual_analysis as Record<string, any>)
        : {};

    const analysisText = String(visual_analysis).toLowerCase();

    // Determine critical_issues_enum
    let criticalIssuesEnum: CriticalIssuesStatus = "none";
    if (analysisDict.critical_issues_enum) {
      criticalIssuesEnum = analysisDict.critical_issues_enum;
    } else if (analysisText.includes("missing sections")) {
      criticalIssuesEnum = "missing_sections";
    } else if (analysisText.includes("critical")) {
      criticalIssuesEnum = "other_issues";
    }

    // Determine visual_changes_enum
    let visualChangesEnum: VisualChangesStatus = "none";
    if (analysisDict.visual_changes_enum) {
      visualChangesEnum = analysisDict.visual_changes_enum;
    } else if (
      analysisText.includes("significant changes") &&
      !analysisText.includes("no significant changes")
    ) {
      visualChangesEnum = "significant";
    } else if (analysisText.includes("minor")) {
      visualChangesEnum = "minor";
    }

    // Determine recommendation_enum
    let recommendationEnum: RecommendationStatus = "pass";
    if (analysisDict.recommendation_enum) {
      recommendationEnum = analysisDict.recommendation_enum;
    } else if (analysisText.includes("review required")) {
      recommendationEnum = "review_required";
    }

    // Create page report with guaranteed valid status
    const pageReport: PageReport = {
      page_path: page_path,
      url: base_url,
      preview_url: pr_url,
      status: status, // Use our determined status instead of parsed one
      critical_issues: parsedReport.critical_issues || {
        sections: [],
        summary: "",
      },
      critical_issues_enum: criticalIssuesEnum,
      structural_analysis: parsedReport.structural_analysis || {
        section_order: "",
        layout: "",
        broken_layouts: "",
      },
      visual_changes: parsedReport.visual_changes || {
        diff_highlights: [],
        animation_issues: "",
        conclusion: "",
      },
      visual_changes_enum: visualChangesEnum,
      recommendation_enum: recommendationEnum,
      conclusion: parsedReport.conclusion || {
        critical_issues: "",
        visual_changes: "",
        recommendation: "",
        summary: "",
      },
      image_refs: image_refs || null,
    };

    reports.push(pageReport);
  }

  return {
    test_data: testData,
    reports: reports,
  };
}
